torchqdynamics/mesolve.py

```diff
@@ -1,27 +1,90 @@
-from typing import Any, Dict, List, Optional
+from typing import List, Tuple
 
 import numpy as np
 import torch
-import torchdiffeq
+from tqdm import tqdm
+
+from torchqdynamics import hermitian
 
 
 def mesolve(
     H: torch.Tensor,
-    Ls: List[torch.Tensor],
     rho0: torch.Tensor,
-    tlist: np.ndarray,
-    solver: str = 'rk4',
-    options: Optional[Dict[str, Any]] = None,
-) -> torch.Tensor:
-    if options is None:
-        options = {}
+    times: np.ndarray,
+    c_ops: List[torch.Tensor] = None,
+    e_ops: List[torch.Tensor] = None,
+    n_substeps: int = 1,
+    save_states=True,
+    progress_bar=False,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    assert n_substeps > 0
+    c_ops = c_ops or []
+    e_ops = e_ops or []
+
+    dt = (times[1] - times[0]) / n_substeps
+    c_dag_c = torch.stack([hermitian(op) @ op for op in c_ops]).sum(dim=0)
+    R = (
+        torch.eye(*H.shape)
+        + (-1j * H + 0.5 * c_dag_c) @ (1j * H + 0.5 * c_dag_c) * dt**2
+    )
+    inv_sqrt_R = _inv_sqrtm(R)
+
+    M_r = torch.eye(*H.shape, dtype=H.dtype) - (1j * H + 0.5 * c_dag_c) * dt
+    M_r_tilde = M_r @ inv_sqrt_R
+
+    c_ops_tilde = [op @ inv_sqrt_R for op in c_ops]
+    e_ops = torch.stack(e_ops)
+
+    rho = torch.clone(rho0)
+    states, measures = [], []
+
+    if progress_bar:
+        times = tqdm(times)
+
+    for _ in times:
+        for _ in range(n_substeps):
+            next_rho = M_r_tilde @ rho
+            next_rho = next_rho @ hermitian(M_r_tilde)
+            next_rho += sum(
+                [
+                    c_op_tilde @ rho @ hermitian(c_op_tilde) * dt
+                    for c_op_tilde in c_ops_tilde
+                ]
+            )
```

Comment:
There's a specific speedup for time-independent Hamiltonian that we could use, where one pre-computes the action of $e^{-iH\Delta t/2}$ and applies it both on the right and on the left. We can keep this for a later PR if you want.

Reply:
See also Pierre's last paper section 3.4 for this (2022: [A tutorial introduction to quantum stochastic master equations based on the qubit/photon system](https://www.sciencedirect.com/science/article/abs/pii/S1367578822001006) section 3.4. Kraus maps and numerical schemes for diffusive SME).